### PR TITLE
Add partition key configuration description

### PR DESCRIPTION
**Pull request by moarychan**
Related PR: [repair partition selection](https://github.com/Azure/azure-sdk-for-java/pull/19787)

Add description for partition key configuration.

### CODE REVIEW

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 1]
```diff
@@ -63,59 +65,78 @@ Name | Description | Required | Default
...
+**_send-timeout_**
 
- Default: `10000`
- 
- #### Event Hub Consumer Properties ####
+Effective only if `sync` is set to true. The amount of time to wait for a response from Event Hub after a send operation, in milliseconds.
+
+Default: `10000`
+
+**_partition-count_**
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:83]: is this a producer property? I mean, is this property `spring.cloud.stream.eventhub.bindings.<channelName>.producer.partition-coung`?

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 2]
```diff
@@ -38,10 +38,12 @@ that's why you have to fill `spring.cloud.stream.eventhub.checkpoint-storage-acc
...
-Event Hub provides similar concept of physical partition as Kafka. But unlike Kafka's auto rebalancing between 
-consumers and partitions, Event Hub provides a kind of preemptive mode. Storage account acts as lease to 
-determine which partition is owned by which consumer. When a new consumer starts, it will try to steal some partitions 
-from most heavy-loaded consumer to achieve workload balancing.
+Event Hub provides a similar concept of physical partition as Kafka. But unlike Kafka's auto rebalancing between consumers and partitions, Event Hub provides a kind of preemptive mode. The storage account acts as a lease to determine which partition is owned by which consumer. When a new consumer starts, it will try to steal some partitions
+from most heavy-loaded consumers to achieve the workload balancing.
+
+The default partition key uses the hash code value of byte[] type `Message.payload`.
+You can use the configuration of Producer `partition-key-extractor-name`, or `partition-key-expression` to specify
+how to obtain the partition key to calculate the partition ID.
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:46]: The last paragraph talks about the partition concept of Event Hubs, but this paragraph talks about Spring Cloud Stream implementation. So perhaps we could delete this directly.

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 3]
```diff
@@ -63,59 +65,78 @@ Name | Description | Required | Default
...
+
+Default: `10000`
+
+**_partition-count_**
+
+Optional. Effective only if `partition-key-expression` or `partition-key-extractor-name` is set. The partition size of Event Hub.
+
+Default: 1
+
+**_partition-key-extractor-name_**
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:89]: same here

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 4]
```diff
@@ -63,59 +65,78 @@ Name | Description | Required | Default
...
+
+Optional. It is the name of a Bean, which implements interface `PartitionKeyExtractorStrategy`.
+`PartitionHandler#extractKey`, the partition handler will first use the `PartitionKeyExtractorStrategy#extractKey` method to apply the message to be sent to obtain the partition key value.
+
+Default: N/A
+
+**_partition-key-expression_**
+
+Optional. It is a SpEL expression that only applies to the message to be sent. When interface `PartitionKeyExtractorStrategy#extractKey` is not implemented, it will be called in the method `PartitionHandler#extractKey`.
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:98]: Same here. And how does the Spring Cloud Stream doc describe this property?

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 5]
```diff
@@ -63,59 +65,78 @@ Name | Description | Required | Default
...
 
-  **_start-position_**
+`BATCH`, checkpoints occur after each batch of messages successfully processed by user-defined message handler without any exception. `default` mode. You may experience reprocessing at most one batch of messages when message processing fails. Be aware that batch size could be any value.
 
-  Whether the consumer receives messages from the beginning or end of event hub. if `EARLIEST`, from beginning. If 
-  `LATEST`, from end.
+`MANUAL`, checkpoints occur on demand by the user via the `Checkpointer`. You can do checkpoints after the message has been successfully processed. `Message.getHeaders.get(AzureHeaders.CHECKPOINTER)`callback can get you the `Checkpointer` you need. Please be aware all messages in the corresponding Event Hub partition before this message will be considered as successfully processed.
 
-  Default: `LATEST`
+`PARTITION_COUNT`, checkpoints occur after the count of messages defined by `checkpoint_count` successfully processed for each partition. You may experience reprocessing at most `checkpoint_count` of  when message processing fails.
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:123]: please help check do we still have this check-point mode?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md (hunk 1 above)
Comment (moarychan) [reply sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:83]: To be precise, it should be `spring.cloud.stream.bindings.<channelName>.producer.partition-count`.

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md (hunk 5 above)
Comment (moarychan) [reply sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:123]: Currently, the check-point-mode is still available.

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 6]
```diff
@@ -55,67 +53,102 @@ The binder provides the following configuration options in `application.properti
...
- **_sync_**
- 
- Whether the producer should act in a synchronous manner with respect to writing messages into a stream. If true, the 
- producer will wait for a response from Event Hub after a send operation.
+You can use the producer configurations of **Spring Cloud Stream**, 
+it uses the configuration with the format of `spring.cloud.stream.bindings.<channelName>.producer`.
+
+##### Partition configuration
+
+The partition key is the hash value of the message payload byte[] type by default.
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:68]: We could change this description to a picture.

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md [hunk 7]
```diff
@@ -188,3 +221,4 @@ Please follow [instructions here][contributing_md] to build from source or contr
 [contributing_md]: https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/spring/CONTRIBUTING.md
 [azure_event_hub]: https://azure.microsoft.com/services/event-hubs/
 [environment_checklist]: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/spring/ENVIRONMENT_CHECKLIST.md#ready-to-run-checklist
+[spring_cloud_stream_3.1.1_producer_properties]: https://docs.spring.io/spring-cloud-stream/docs/3.1.1/reference/html/spring-cloud-stream.html#_producer_properties
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/README.md:224]: probably change the 3.1.1 to current.

**Review by saragluna — CHANGES_REQUESTED** [review]

**Review by saragluna — APPROVED** [review]
LGTM.

**Pull request merged**